### PR TITLE
Nerfs the AI's tracking ability.

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

The AIs tracking ability now requires a target to stand still for 3 seconds in order to acquire a track. If the target moves during this time, the track will fail.
You can instantly track a target by double clicking them.
You can instantly track anyone with their suit sensor tracking beacon enabled.
You can instantly track silicons.
If you have malf upgrades, you can instantly track.

## Why It's Good For The Game

The AI can no longer instantly locate someone on the station that is running around.
The ability to instantly track anything near a camera is incredibly powerful, and requires almost no thought to use. The ability for any member of security to say "AI locate [X]" and instantly get a track on someone on cameras is incredibly powerful.
If you locate a target visually, you can double click them to acquire a track instantly but if you want to locate someone on the entire map, you need them to stand still for a short period of time first. This means that someone running away from security may be able to enter maintenance to get away from the cameras and can survive without instantly being retracked.
The main positive uses for this ability would be to recover bodies, however since bodies aren't moving, if they are on the camera network you are still able to locate them; plus any good citizens with their sensors on will be able to be tracked within an instant. I intent to nerf the effectiveness of this ability against the not so good citizens that are likely to be hiding from security with their sensors disabled.

## Testing Photographs and Procedure

image

image

## Changelog
:cl:
balance: AI tracking now requires a target to stand still for X seconds in order to acquire a track when used from the menu or from radio. X is calculated using the following formula. If the target has GPS tracking beacon, X is 0. If not X starts at 4. Subtract 1 second if the target is in the light. Subtract 1 second if the target's ID matches their name. Subtract 1 second if the target's face is visible and not disfigured.
balance: AI tracking can instantly track a target if they have sensors enabled, or are a robot.
/:cl:

### CODE REVIEW

**Comment by itsmeow** [comment]
I am happy with this change, good for the game

**Review by itsmeow — DISMISSED** [review]

**Comment by FortSumter** [comment]
I deleted an essay on why this change is bad. Instead I would like to just ask you to not fuck with the AI.

I get that this is a problem related to game balance you're trying to fix, but all you're doing is discouraging people from playing the role at all. what I do all shift as AI is jump between people talking over the radio channels to see what they're doing. You will essentially be taking away my ability to enjoy the role by making it so I can't reliably do that. There are so many other aspects of the AI that need rebalancing or updating and you choose the worst one to PR.

You are also encouraging the mandatory use of maxed suit sensors. I think max suit sensors is too powerful with too few downsides or counters. Suit sensors is meant to be a choice and making the AI unable to help you if you don't max them is kind of shitty.

**Comment by PowerfulBacon** [comment]
It's 3 seconds, anyone asking for a door to be opened will be standing still for that time, meaning you will still be able to jump to them and help them.
Not being able to instantly see what anyone is doing at any time by teleporting across the map instantly is the problem and the reason why AIs tend to make the game significantly less fun for other players.
If you have suit sensors to maximum then your position can be located at all times and I think the other nerfs to it are sufficient enough to reduce its power to counter combat and screw with antagonists trying to get stealthy kills.

**Comment by FortSumter** [comment]
It still doesn't allow antags to get stealthy kills though. If the target yells over coms with maxed sensors, the antags caught. 3 seconds is absurdly long, especially when server dilation can make 3 seconds 1 minute. All your doing is making the AI useless and making maxed suit sensors mandatory. It's meant to be the most advanced thing on the station and yet most people only use it to open doors and sometimes I feel that's all I can do as AI.

**Comment by Oscari-pog** [comment]
Good change. Dead people wont be doing much moving and people waiting for doors stay still too.

**Comment by llol111** [comment]
i dont like this.
Thats all.  Feels like nerf just to be nerfing something.

**Comment by PowerfulBacon** [comment]
> i dont like this.
> Thats all.  Feels like nerf just to be nerfing something.

The reason for changes are in the PR body

**Comment by llol111** [comment]
Do you mean 
>"The ability to instantly track anything near a camera is incredibly powerful, and requires almost no thought to use. The ability for any member of security to say "AI locate [X]" and instantly get a track on someone on cameras is incredibly powerful."
>"This means that someone running away from security may be able to enter maintenance to get away from the cameras and can survive without instantly being retracked."
?

It seems in the massive minority here but i dont think thats a bad thing. Its intented to be this way and has be fine for year this way. If an antag does not want to be tracked they can buy an agent ID.

**Comment by PowerfulBacon** [comment]
Balancing all gamemodes and antagonists around a single one is not a good way to go, only traitors can buy the agent card and it sucks to tell someone that they have to buy 20 TC worth of items to have immunity to all of the things that are too powerful to begin with.

**Comment by Sinestia** [comment]
> Do you mean
> 
> > "The ability to instantly track anything near a camera is incredibly powerful, and requires almost no thought to use. The ability for any member of security to say "AI locate [X]" and instantly get a track on someone on cameras is incredibly powerful."
> > "This means that someone running away from security may be able to enter maintenance to get away from the cameras and can survive without instantly being retracked."
> > ?
> 
> It seems i am in the massive minority here but i dont think thats a bad thing. Its intented to be this way and has be fine for years this way. If an antag does not want to be tracked they can buy an agent ID.

Just because it's been that way for years doesn't mean it's necessarily fine 

Sometimes shitty ideas stay around

**Comment by FortSumter** [comment]
> Sometimes shitty ideas stay around

And some shitty ideas show up here. 

The problem is how to keep the AI from instantly hunting down people using the tracker without hindering the AIs ability to function efficiently.

The AI is meant to be an all powerful, all seeing entity; the eye in the sky, the ghost in the machine, the man in the walls, ect. If this is really how bacon wants to solve this problem, I want something to balance it in return. 

The menu the AI uses to jump to crew is criminally outdated and only updates if you close it and re-open. I'd like to see the menu get a fresh coat of paint and automatically update with the names of crew members as they become available to jump to. That way this change would be more manageable while also giving the AI player the intended feeling of being omniscient within reason. Once that was done, you could play around with the jump delay and see if it should be longer or shorter.

**Comment by hotimep** [comment]
i will have an excellent time telling security ;sorry i can't track the guy mass murdering people they haven't been standing still for 3 seconds

**Comment by PowerfulBacon** [comment]
The idea that powerful things shouldn't be made less powerful but replaced with other equally powerful things is a fundamentally flawed approach to looking at game design.
While managing the people on the station is one of the jobs on AI, being able to instantly locate anyone on the station kills the need to think about where to search when trying to hunt someone down and almost kills the time needed to find antagonists, which leads to situations being resolved too quickly, security being bored and ultimately more boring gameplay and rounds.

**Comment by FortSumter** [comment]
Perhaps the most simple solution would be to remove the efficiency and corporate lawsets and add RP lawsets so the AI can't just validhunt for security. The efficiency lawset in particular has been a problem since inception, as it's worse then just purging it imo.

**Comment by Tsar-Salat** [comment]
> Ps have you ever tried double clicking a player running around? I have hundreds of hours as AI and I can rarely do it.

yeah, this. I am a silicon main and I constantly need to rely on radio, trackers and sensors to follow people around. The janky-ass camera simply cannot keep up with run speeds like trackers can. I feel like the intent of this PR is good, but its core design is somewhat damaged by what I feel is lack of experience in the role

**Comment by PowerfulBacon** [comment]
This PR doesn't remove the ability to have the camera live follow someone while they are moving, once you have aquired the track, the track stays. It makes it so you cannot aquire a track on a moving target. (I'd be fine with allowing you to instantly track people in view of the camera, but people across the map shouldnt)

**Comment by Howluinb** [comment]
Honestly, this is a good change. 3 seconds is really long, though, but 2 seems more fair and better for fluidity's sake. Maybe 1.5 ?
That delay is however really good tho. Not everyone has the ability to buy an agent ID - and balancing everything around tot sucks ass and is how we got the shotgun meta in the first place.

**Comment by PowerfulBacon** [comment]
I am happy to make it so the AI can instantly track anyone within view of their current eye, but across the map should be slow and less viable.

**Comment by MCterra10** [comment]
Have people really forgotten the two steps it takes to disable a camera? lol
Ided PR

File: code/game/machinery/camera/tracking.dm [hunk 1]
```diff
@@ -67,8 +67,36 @@
...
+	attempt_track(target)
+
+/mob/living/silicon/ai/proc/attempt_track(mob/living/target)
+	//If the target has sensors on, track instantly
+	//If the AI has malf upgrades, allow instant tracking
+	var/instant_track = !!malf_picker || issilicon(target) || isbot(target)
+
+	if(!instant_track && isliving(target))
+		var/mob/living/L = target
+		if(HAS_TRAIT(L, TRAIT_SUIT_SENSORS) || HAS_TRAIT(L, TRAIT_NANITE_SENSORS))
```
Comment (PowerfulBacon) [inline code/game/machinery/camera/tracking.dm]: This works even if the location beacon is disabled

Comment (PowerfulBacon) [reply code/game/machinery/camera/tracking.dm]: This is fine actually, it only really benefits you then if you actively disable your sensors so has less impact on the AI

**Comment by Shiraizawa** [comment]
Can't you just look someone up from the camera selector dropdown list thing (that ai ability) and then just double click them to get a nearly-instantaneous track?

**Comment by PowerfulBacon** [comment]
Thats the specific thing that this PR counters, looking up crew members from the track list that don't have any sensors on. You won't notice this at all if you aren't using that menu since you can't see crew members with their sensors disabled in the crew monitoring menu. In game this has very little noticable impact within the normal flow of a round

**Comment by Shiraizawa** [comment]
> Thats the specific thing that this PR counters, looking up crew members from the track list that don't have any sensors on. You won't notice this at all if you aren't using that menu since you can't see crew members with their sensors disabled in the crew monitoring menu. In game this has very little noticable impact within the normal flow of a round

You absolutely could do that before. Im not talking about the suit sensors network, but the one which you can see any animal/monkeys/humans that appear on your cameras. Just get the dropdown window up, hit the first character of their name, that jumps close enough to them, then press enter .

**Comment by PowerfulBacon** [comment]
You can still do that, but if the target has no sensors then it takes time to track them

File: code/game/machinery/camera/tracking.dm [hunk 2]
```diff
@@ -67,8 +67,36 @@
...
+			instant_track = TRUE
+
+	//Require the target to remain still for 3 seconds in order to acquire the track.
+	//Once track is acquired, it will hold and follow them while moving
+	if(!instant_track)
+		to_chat(src, "<span class='notice'>Target not present on suit sensor network, querying facial recognition network...</span>")
```
Comment (PowerfulBacon) [inline code/game/machinery/camera/tracking.dm]:
```suggestion
		to_chat(src, "<span class='notice'>Target has no suit sensor beacon, querying facial recognition network. Please wait...</span>")
```

**Comment by PowerfulBacon** [comment]
image
Retested

**Review by itsmeow — APPROVED** [review]

**Pull request merged**